feat(spanner): add ToStructLenient method to decode to struct fields with no error return in case of un-matched row's column with struct's exported fields

rahul2393 opened the pull request:
The PR adds method `ToStructLenient` which decodes the row's column into the struct's exported fields, the difference with the existing `ToStruct` methods are:
* Decoding from row's column to struct with no matching exported fields is allowed.

ToStructLenient should pass all existing and future test cases for ToStruct method so tests are updated to make sure same test cases run for both methods.

ToStructLenient is created to handle below example:

DDL:
```
CREATE TABLE Singers (
    SingerId	INT64 NOT NULL,
    FirstName	STRING(1024)
) PRIMARY KEY (SingerId)
```
and client code looks like
```
struct Singer {
  SingerId     int64
  FirstName string
}

func GetSinger(ID int64) (*Singer, error) {
  ...
  spannerCli.Single().Query(ctx, spanner.Statement {
      SQL: 'SELECT * FROM singers WHERE SingerId=@id',
      Params: map[string]string{"id": ID},
  ).Do(func(r *spanner.Row) error {
       singer := new(Singer)
       r.ToStruct(singer)
  )
}
```

Now, let's say we update the singers table to have new column `LastName`, ToStruct will start returning errNoOrDupGoField after schema update to fix this ToStructLenient can be used instead of ToStruct.
  
Fixes: https://github.com/googleapis/google-cloud-go/issues/5131

git-hulk commented:
Cool, many thanks to @rahul2393

olavloite commented on spanner/value.go:
```diff
@@ -2871,6 +2871,11 @@ func errNilSpannerStructType() error {
 
+// errDupGoField returns error for duplicated Go STRUCT field names
+func errDupGoField(s interface{}) error {
+	return spannerErrorf(codes.InvalidArgument, "Go struct %+v(type %T) has duplicate fields for GO STRUCT field", s, s)
```
nit: Could this error also include the field that is duplicate? I think that would be helpful to any user who needs to debug code that returns this error.

rahul2393 replied:
Done

olavloite commented on spanner/value.go:
```diff
@@ -2921,14 +2926,21 @@ func decodeStruct(ty *sppb.StructType, pb *proto3.ListValue, ptr interface{}) er
+	for _, f := range fieldNames {
+		sf := fields.Match(f)
+		if sf == nil {
+			return errDupGoField(ptr)
```
Is this really a duplicate field? Or is it a missing field? I would expect `fields.Match(f)` to return nil if the field name is not in the list, and that this scenario is that the field is missing and not duplicate.

rahul2393 replied:
Here it is duplicate field, it would be missing if we would have tried matching the spanner.Row field name, but here the same GO struct field names are matched from which the fields object was created.

olavloite commented on spanner/value.go:
```diff
@@ -2995,6 +3007,38 @@ func decodeStructArray(ty *sppb.StructType, pb *proto3.ListValue, ptr interface{
 
+func getAllFieldNames(v reflect.Value) []string {
+	var names []string
+	s := v
```
Why do we need this?

rahul2393 replied:
We need this for cases when GO struct has more than one common spanner tags, In that case we want to return duplicate error. There was no way to get all exported struct field names so created one here. please let me know if you think there is another good way to do this.

olavloite replied:
Sorry, I should have been more specific. I meant literally why do we need this specific line:

`s := v`

Why can't the method just use `v`?

olavloite reviewed (commented):
There seems to be some integration tests that are failing. Would you mind taking a look at those?

olavloite commented:
> Fixes: #5131

Also, would you mind extending the PR description to describe what the PR does. I understand that it solves the issue, but adding a short description of what/how makes it easier to review and will automatically be included in the release notes.

rahul2393 commented:
Thanks @olavloite, Integreation tests passed and updated the PR description.

olavloite commented:
@rahul2393 Thanks for the changes. The implementation itself looks good to me.

I however have a couple of general concerns / questions regarding this:
1. First of all, I think this PR should be labelled `feat` and not `fix`. This is not fixing a bug, but relaxing a restriction that appears to have been added quite deliberately when the feature was created. Would you mind extending the PR description a little bit more to explain what is possible now that was not possible before (preferably with an example)?
2. I'm also wondering whether we need to mark this as a breaking change, or otherwise at least flag it more prominently in the release notes than a 'normal' feature, as we are changing the behavior of the method. The `ToStruct` method would previously return an error in certain cases, and after this change, it will stop returning an error. In theory, there could be users who rely on an error being returned in those cases, and whose code would break after this change.
3. We should update the documentation of the `ToStruct` [method here](https://github.com/googleapis/google-cloud-go/blob/90e2868a3d220aa7f897438f4917013fda7a7c59/spanner/row.go#L280). As far as I can tell, this change does not contradict with anything in the documentation of that method, but that is because the documentation is not explicit in what happens if the result set contains more/less columns than there are fields in the struct. The documentation should be extended to explicitly state what the method does if the exported fields of the struct and the columns in the result set do not match.
4. The documentation also does not explicitly state that it will only fill exported fields. From what I can tell, both the current implementation and the one in this PR will only fill exported fields, and I think the documentation should reflect that.

rahul2393 commented:
@olavloite I have updated the PR to add new method instead of updating the existing ToStruct method so it won't be a breaking change, also updated the PR description. same tests should be run for ToStructLenient which were created for ToStruct so tests are updated, this will make sure that functionality wise ToStruct and ToStructLenient works same with just one difference of relaxing restriction on matching row's column with struct exported field.
Please help review again.

olavloite commented on spanner/row.go:
```diff
@@ -311,5 +312,43 @@ func (r *Row) ToStruct(p interface{}) error {
+// Slice and pointer fields will be set to nil if the source column is NULL, and a
+// non-nil value if the column is not NULL. To decode NULL values of other types, use
+// one of the spanner.NullXXX types as the type of the destination field.
+func (r *Row) ToStructLenient(p interface{}) error {
```
```suggestion
// If ToStruct returns an error, the contents of p are undefined. Some fields may
// have been successfully populated, while others were not; you should not use any of
// the fields.
func (r *Row) ToStructLenient(p interface{}) error {
```

This warning is still true for this method. Although this method is less likely to return an error, it can still fail if for example the type of one of the fields is not compatible with the datatype of the corresponding column.

olavloite commented on spanner/row.go:
```diff
@@ -311,5 +312,43 @@ func (r *Row) ToStruct(p interface{}) error {
+//
+//   2. Otherwise, if the name of a field matches the name of a column (ignoring case),
+//      decode the column into the field.
+//
```
```suggestion
//
// 3. The number of columns in the row and exported fields in the struct do not need to match.
// Any field in the struct that cannot not be assigned a value from the row is assigned its default value.
// Any column in the row that does not have a corresponding field in the struct is ignored.
```

olavloite commented on spanner/value.go:
```diff
@@ -2986,7 +3003,7 @@ func decodeStructArray(ty *sppb.StructType, pb *proto3.ListValue, ptr interface{
 		}
 		// Decode proto3.ListValue l into struct referenced by s.Interface().
-		if err = decodeStruct(ty, l, s.Interface()); err != nil {
+		if err = decodeStruct(ty, l, s.Interface(), false); err != nil {
```
Shouldn't this be:
```suggestion
		if err = decodeStruct(ty, l, s.Interface(), lenient); err != nil {
```

rahul2393 replied:
No, this is called from decodeStructArray and we don't want to change that function.

olavloite replied:
Ah, yeah, good point. I missed that.

olavloite commented on spanner/value.go:
```diff
@@ -2921,13 +2926,25 @@ func decodeStruct(ty *sppb.StructType, pb *proto3.ListValue, ptr interface{}) er
 	if err != nil {
 		return ToSpannerError(err)
 	}
+	// return error if linent is true and destination has duplicate exported columns
```
nit:
```suggestion
	// return error if lenient is true and destination has duplicate exported columns
```

rahul2393 commented on spanner/row.go:
```diff
@@ -295,7 +295,8 @@ func errToStructArgType(p interface{}) error {
 //
-// If ToStruct returns an error, the contents of p are undefined. Some fields may
+// If ToStruct returns an error, the contents of p are undefined or if the there is un-matching
+// column from row's columns into struct's exported fields. Some fields may
```
> if an exported field in the struct cannot not be assigned a value from the row.
I think this is not the case with ToStruct now, it won't return error.

hengfengli commented on spanner/value.go:
```diff
@@ -2921,13 +2926,25 @@ func decodeStruct(ty *sppb.StructType, pb *proto3.ListValue, ptr interface{}) er
 	if err != nil {
 		return ToSpannerError(err)
 	}
+	// return error if lenient is true and destination has duplicate exported columns
```
nit: return -> Return
